### PR TITLE
fastfield return iterator instead fill vec

### PR DESCRIPTION
**Pull request by PSeitz**
return iterator from get_vals method. This will allow to save on unnecessary vec allocations.

### CODE REVIEW

File: fastfield_codecs/src/blockwise_linear.rs [hunk 1]
```diff
@@ -300,13 +300,9 @@ impl FastFieldCodec for BlockwiseLinearCodec {
...
             fastfield_accessor.max_value() - fastfield_accessor.min_value();
-        if fastfield_accessor
+        fastfield_accessor
             .max_value()
-            .checked_add(theorethical_maximum_offset)
-            .is_none()
-        {
-            return None;
-        }
+            .checked_add(theorethical_maximum_offset)?;
```
Comment (fulmicoton) [inline fastfield_codecs/src/blockwise_linear.rs:305]: I prefer the earlier version... By far actually.

In the new version the intent is hidden.

File: fastfield_codecs/src/column.rs [hunk 2]
```diff
@@ -8,23 +10,14 @@ pub trait Column<T = u64> {
...
-    /// Regardless of the type of `Item`, this method works
-    /// - transmuting the output array
-    /// - extracting the `Item`s as if they were `u64`
-    /// - possibly converting the `u64` value to the right type.
+    /// Returns an iterator over given doc range.
     ///
     /// # Panics
     ///
-    /// May panic if `start + output.len()` is greater than
+    /// May panic if `range.end()` is greater than
```
Comment (fulmicoton) [inline fastfield_codecs/src/column.rs:19]:
```suggestion
    /// May panic if `range.end` is greater than
```

File: src/fastfield/facet_reader.rs [hunk 3]
```diff
@@ -76,7 +76,8 @@ impl FacetReader {
 
     /// Return the list of facet ordinals associated to a document.
     pub fn facet_ords(&self, doc: DocId, output: &mut Vec<u64>) {
-        self.term_ords.get_vals(doc, output);
```
Comment (fulmicoton) [inline src/fastfield/facet_reader.rs:79]: so here, this is a net loss.

File: fastfield_codecs/src/linear.rs [hunk 4]
```diff
@@ -204,13 +204,9 @@ impl FastFieldCodec for LinearCodec {
...
             fastfield_accessor.max_value() - fastfield_accessor.min_value();
-        if fastfield_accessor
+        fastfield_accessor
             .max_value()
-            .checked_add(theorethical_maximum_offset)
-            .is_none()
-        {
-            return None;
-        }
+            .checked_add(theorethical_maximum_offset)?;
```
Comment (fulmicoton) [inline fastfield_codecs/src/linear.rs:209]: same as before. I prefer the earlier version.

**Comment by fulmicoton** [comment]
Do you have benchmark to confirm we don't have a regression in search?

The `fill` to buffer indirection point is that it get rids of dynamic dispatch.

File: fastfield_codecs/src/blockwise_linear.rs (hunk 1 above)
Comment (PSeitz) [reply fastfield_codecs/src/blockwise_linear.rs:305]: it's from a clippt lint, I also think the previous version is more clear. We could use [overflowing_add](https://doc.rust-lang.org/std/primitive.i8.html#method.overflowing_add) instead and check the bool, or disable the lint

Comment (fulmicoton) [reply fastfield_codecs/src/blockwise_linear.rs:305]: What about disabling the lint locally?

**Comment by PSeitz** [comment]
> Do you have benchmark to confirm we don't have a regression in search?
> 
> The `fill` to buffer indirection point is that it get rids of dynamic dispatch.

I removed the dynamic dispatch and added a bench. It is still slower, which I don't understand yet. 

```
Vec version - Most time spent in MultiValuedFastFieldReader<Item>::get_vals
running 1 test
test fastfield::bench::bench_multi_value_fflookup                             ... bench:     758,148 ns/iter (+/- 13,857)

+   93,36%    15,12%  tantivy-cbbeafa  tantivy-cbbeafab3d8c2ac9      [.] test::bench::Bencher::iter                                                                ▒
+   78,22%    78,20%  tantivy-cbbeafa  tantivy-cbbeafab3d8c2ac9      [.] tantivy::fastfield::multivalued::reader::MultiValuedFastFieldReader<Item>::get_vals       ▒


Iter version Static dispatch - Most time spent in sum
running 1 test
test fastfield::bench::bench_multi_value_fflookup                             ... bench:   1,382,918 ns/iter (+/- 32,6

+   96,21%    32,31%  tantivy-ea620f5  tantivy-ea620f5d7d4bf2df      [.] test::bench::Bencher::iter                                                               ▒
+   50,75%    50,73%  tantivy-ea620f5  tantivy-ea620f5d7d4bf2df      [.] <u64 as core::iter::traits::accum::Sum>::sum                                             ▒
+   13,12%    13,10%  tantivy-ea620f5  tantivy-ea620f5d7d4bf2df      [.] tantivy::fastfield::multivalued::reader::MultiValuedFastFieldReader<Item>::get_vals      ▒



Iter version dynamic dispatch
running 1 test
test fastfield::bench::bench_multi_value_fflookup                             ... bench:   1,550,563 ns/iter (+/- 31,509

```

**Comment by fulmicoton** [comment]
@PSeitz Let's stick to the Vec filling solution then. It is much easier to reason with and allocation is usually not an issue (we reuse the same buffer over and over).

By using the iterator solution we are making the job of the CPU and the compiler very hard.

**Comment by fulmicoton** [comment]
@PSeitz Closing this. (Feel free to reopen if I missed something.)

**Pull request closed without merging**